File: services/mcp/slashAddPlan.ts
import { randomBytes } from 'crypto'
import { addMcpConfig } from './config.js'
import { parseHeaders } from './utils.js'
import { parseEnvVars } from '../../utils/envUtils.js'
import {
  McpServerConfigSchema,
  type ConfigScope,
  type McpServerConfig,
} from './types.js'

export const MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS = 10 * 60 * 1000

export type McpSlashAddWritableScope = Extract<
  ConfigScope,
  'local' | 'user' | 'project'
>

export type McpSlashAddTransport = 'stdio' | 'sse' | 'http'

export type McpSlashAddPlan = {
  token: string
  createdAt: number
  serverName: string
  scope: McpSlashAddWritableScope
  transport: McpSlashAddTransport
  config: McpServerConfig
}

export type McpSlashAddPlanError =
  | {
      type: 'missing_server_name'
    }
  | {
      type: 'missing_command'
    }
  | {
      type: 'invalid_scope'
      scope?: string
    }
  | {
      type: 'invalid_transport'
      transport?: string
    }
  | {
      type: 'invalid_env'
      message: string
    }
  | {
      type: 'invalid_header'
      message: string
    }
  | {
      type: 'invalid_config'
      reason: string
    }
  | {
      type: 'unknown_token'
      token: string
    }
  | {
      type: 'expired_token'
      token: string
    }
  | {
      type: 'install_failed'
      message: string
    }

export type McpSlashAddPlanResult =
  | {
      ok: true
      plan: McpSlashAddPlan
    }
  | {
      ok: false
      error: McpSlashAddPlanError
    }

const planStore = new Map<string, McpSlashAddPlan>()

function pruneExpiredPlans(now = Date.now()): void {
  for (const [token, plan] of planStore.entries()) {
    if (now - plan.createdAt > MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS) {
      planStore.delete(token)
    }
  }
}

function createToken(): string {
  let token = randomBytes(4).toString('hex')
  while (planStore.has(token)) {
    token = randomBytes(4).toString('hex')
  }
  return token
}

function normalizeScope(scope?: string): McpSlashAddWritableScope | null {
  if (!scope || scope === 'local') return 'local'
  if (scope === 'user' || scope === 'project') return scope
  return null
}

function normalizeTransport(transport?: string): McpSlashAddTransport | null {
  if (!transport || transport === 'stdio') return 'stdio'
  if (transport === 'sse' || transport === 'http') return transport
  return null
}

function formatSchemaError(error: unknown): string {
  const issues =
    (error as {
      issues?: Array<{
        path: Array<string | number>
        message: string
      }>
    }).issues ?? []
  return issues
    .map(issue => `${issue.path.join('.')}: ${issue.message}`)
    .join(', ')
}

export function getMcpSlashAddPlan(opts: {
  serverName?: string
  scope?: string
  transport?: string
  commandOrUrl?: string
  args?: string[]
  env?: string[]
  headers?: string[]
}): McpSlashAddPlanResult {
  pruneExpiredPlans()

  const serverName = opts.serverName?.trim()
  if (!serverName) {
    return { ok: false, error: { type: 'missing_server_name' } }
  }

  const scope = normalizeScope(opts.scope)
  if (!scope) {
    return { ok: false, error: { type: 'invalid_scope', scope: opts.scope } }
  }

  const transport = normalizeTransport(opts.transport)
  if (!transport) {
    return {
      ok: false,
      error: { type: 'invalid_transport', transport: opts.transport },
    }
  }

  const commandOrUrl = opts.commandOrUrl?.trim()
  if (!commandOrUrl) {
    return { ok: false, error: { type: 'missing_command' } }
  }

  let config: McpServerConfig
  if (transport === 'stdio') {
    let env: Record<string, string> | undefined
    try {
      env = parseEnvVars(opts.env)
    } catch (error) {
      return {
        ok: false,
        error: {
          type: 'invalid_env',
          message: error instanceof Error ? error.message : String(error),
        },
      }
    }
    config = {
      type: 'stdio',
      command: commandOrUrl,
      args: opts.args ?? [],
      ...(Object.keys(env).length > 0 ? { env } : {}),
    }
  } else {
    let headers: Record<string, string> | undefined
    if (opts.headers?.length) {
      try {
        headers = parseHeaders(opts.headers)
      } catch (error) {
        return {
          ok: false,
          error: {
            type: 'invalid_header',
            message: error instanceof Error ? error.message : String(error),
          },
        }
      }
    }
    config = {
      type: transport,
      url: commandOrUrl,
      ...(headers ? { headers } : {}),
    }
  }

  const parsed = McpServerConfigSchema().safeParse(config)
  if (!parsed.success) {
    return {
      ok: false,
      error: {
        type: 'invalid_config',
        reason: formatSchemaError(parsed.error),
      },
    }
  }

  const token = createToken()
  const plan: McpSlashAddPlan = {
    token,
    createdAt: Date.now(),
    serverName,
    scope,
    transport,
    config: parsed.data,
  }
  planStore.set(token, plan)
  return { ok: true, plan }
}

export async function executeMcpSlashAddPlan(
  token: string,
): Promise<McpSlashAddPlanResult> {
  pruneExpiredPlans()
  const plan = planStore.get(token)
  if (!plan) {
    return { ok: false, error: { type: 'unknown_token', token } }
  }

  planStore.delete(token)
  if (Date.now() - plan.createdAt > MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS) {
    return { ok: false, error: { type: 'expired_token', token } }
  }

  try {
    await addMcpConfig(plan.serverName, plan.config, plan.scope)
    return { ok: true, plan }
  } catch (error) {
    return {
      ok: false,
      error: {
        type: 'install_failed',
        message: error instanceof Error ? error.message : String(error),
      },
    }
  }
}

export function _resetMcpSlashAddPlanStoreForTesting(): void {
  planStore.clear()
}

File: crates/mossen-agent/src/mcp/slash_add_plan.rs
//! /mcp add slash command plan — staged install with confirmation token.
//!
//! Translates `services/mcp/slashAddPlan.ts`.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use rand::Rng;

use crate::mcp::types::{ConfigScope, McpServerConfig};

pub const MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS: u64 = 10 * 60 * 1000;

pub type McpSlashAddWritableScope = ConfigScope; // local | user | project
pub type McpSlashAddTransport = &'static str; // "stdio" | "sse" | "http"

#[derive(Debug, Clone)]
pub struct McpSlashAddPlan {
    pub token: String,
    pub created_at: u64,
    pub server_name: String,
    pub scope: ConfigScope,
    pub transport: String,
    pub config: McpServerConfig,
}

#[derive(Debug, Clone)]
pub enum McpSlashAddPlanError {
    MissingServerName,
    MissingCommand,
    InvalidScope { scope: Option<String> },
    InvalidTransport { transport: Option<String> },
    InvalidEnv { message: String },
    InvalidHeader { message: String },
    InvalidConfig { reason: String },
    UnknownToken { token: String },
    ExpiredToken { token: String },
    InstallFailed { message: String },
}

pub enum McpSlashAddPlanResult {
    Ok { plan: McpSlashAddPlan },
    Err { error: McpSlashAddPlanError },
}

lazy_static::lazy_static! {
    static ref PLAN_STORE: Mutex<HashMap<String, McpSlashAddPlan>> = Mutex::new(HashMap::new());
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn prune_expired_plans(store: &mut HashMap<String, McpSlashAddPlan>) {
    let now = now_ms();
    store.retain(|_, plan| now - plan.created_at <= MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS);
}

fn create_token(store: &HashMap<String, McpSlashAddPlan>) -> String {
    let mut rng = rand::thread_rng();
    loop {
        let token = format!("{:08x}", rng.gen::<u32>());
        if !store.contains_key(&token) {
            return token;
        }
    }
}

fn normalize_scope(scope: Option<&str>) -> Option<ConfigScope> {
    match scope {
        None | Some("local") => Some(ConfigScope::Local),
        Some("user") => Some(ConfigScope::User),
        Some("project") => Some(ConfigScope::Project),
        _ => None,
    }
}

fn normalize_transport(transport: Option<&str>) -> Option<&'static str> {
    match transport {
        None | Some("stdio") => Some("stdio"),
        Some("sse") => Some("sse"),
        Some("http") => Some("http"),
        _ => None,
    }
}

/// Parse environment variables from "KEY=VALUE" format.
pub fn parse_env_vars(env: Option<&[String]>) -> Result<HashMap<String, String>, String> {
    let mut map = HashMap::new();
    if let Some(vars) = env {
        for var in vars {
            let eq_idx = var
                .find('=')
                .ok_or_else(|| format!("Invalid env format: \"{}\". Expected KEY=VALUE", var))?;
            let key = var[..eq_idx].trim().to_string();
            let value = var[eq_idx + 1..].trim().to_string();
            if key.is_empty() {
                return Err(format!("Invalid env: \"{}\". Key cannot be empty.", var));
            }
            map.insert(key, value);
        }
    }
    Ok(map)
}

/// Parse headers from "Key: Value" format.
pub fn parse_headers(headers: &[String]) -> Result<HashMap<String, String>, String> {
    let mut map = HashMap::new();
    for header in headers {
        let colon_idx = header.find(':').ok_or_else(|| {
            format!(
                "Invalid header format: \"{}\". Expected format: \"Header-Name: value\"",
                header
            )
        })?;
        let key = header[..colon_idx].trim().to_string();
        let value = header[colon_idx + 1..].trim().to_string();
        if key.is_empty() {
            return Err(format!(
                "Invalid header: \"{}\". Header name cannot be empty.",
                header
            ));
        }
        map.insert(key, value);
    }
    Ok(map)
}

/// Create a /mcp add install plan.
pub fn get_mcp_slash_add_plan(
    server_name: Option<&str>,
    scope: Option<&str>,
    transport: Option<&str>,
    command_or_url: Option<&str>,
    args: Option<&[String]>,
    env: Option<&[String]>,
    headers: Option<&[String]>,
) -> McpSlashAddPlanResult {
    let mut store = PLAN_STORE.lock().unwrap();
    prune_expired_plans(&mut store);

    let server_name = match server_name.map(|s| s.trim()) {
        Some(n) if !n.is_empty() => n.to_string(),
        _ => {
            return McpSlashAddPlanResult::Err {
                error: McpSlashAddPlanError::MissingServerName,
            };
        }
    };

    let resolved_scope = match normalize_scope(scope) {
        Some(s) => s,
        None => {
            return McpSlashAddPlanResult::Err {
                error: McpSlashAddPlanError::InvalidScope {
                    scope: scope.map(|s| s.to_string()),
                },
            };
        }
    };

    let resolved_transport = match normalize_transport(transport) {
        Some(t) => t,
        None => {
            return McpSlashAddPlanResult::Err {
                error: McpSlashAddPlanError::InvalidTransport {
                    transport: transport.map(|s| s.to_string()),
                },
            };
        }
    };

    let command_or_url = match command_or_url.map(|s| s.trim()) {
        Some(c) if !c.is_empty() => c.to_string(),
        _ => {
            return McpSlashAddPlanResult::Err {
                error: McpSlashAddPlanError::MissingCommand,
            };
        }
    };

    let config = if resolved_transport == "stdio" {
        let parsed_env = match parse_env_vars(env) {
            Ok(e) => e,
            Err(msg) => {
                return McpSlashAddPlanResult::Err {
                    error: McpSlashAddPlanError::InvalidEnv { message: msg },
                };
            }
        };
        McpServerConfig::Stdio {
            command: command_or_url,
            args: args.map(|a| a.to_vec()).unwrap_or_default(),
            env: if parsed_env.is_empty() {
                None
            } else {
                Some(parsed_env)
            },
            cwd: None,
        }
    } else {
        let parsed_headers = if let Some(h) = headers {
            if h.is_empty() {
                None
            } else {
                match parse_headers(h) {
                    Ok(h) => Some(h),
                    Err(msg) => {
                        return McpSlashAddPlanResult::Err {
                            error: McpSlashAddPlanError::InvalidHeader { message: msg },
                        };
                    }
                }
            }
        } else {
            None
        };

        if resolved_transport == "sse" {
            McpServerConfig::Sse {
                url: command_or_url,
                headers: parsed_headers,
                headers_helper: None,
                oauth: None,
            }
        } else {
            McpServerConfig::Http {
                url: command_or_url,
                headers: parsed_headers,
                headers_helper: None,
                oauth: None,
            }
        }
    };

    let token = create_token(&store);
    let plan = McpSlashAddPlan {
        token: token.clone(),
        created_at: now_ms(),
        server_name,
        scope: resolved_scope,
        transport: resolved_transport.to_string(),
        config,
    };
    store.insert(token, plan.clone());
    McpSlashAddPlanResult::Ok { plan }
}

/// Execute a previously-created slash add plan.
pub async fn execute_mcp_slash_add_plan(
    token: &str,
    add_mcp_config: &dyn super::builtin_template_plan::AsyncAddMcpConfig,
) -> McpSlashAddPlanResult {
    let plan = {
        let mut store = PLAN_STORE.lock().unwrap();
        prune_expired_plans(&mut store);
        store.remove(token)
    };

    let plan = match plan {
        Some(p) => p,
        None => {
            return McpSlashAddPlanResult::Err {
                error: McpSlashAddPlanError::UnknownToken {
                    token: token.to_string(),
                },
            };
        }
    };

    if now_ms() - plan.created_at > MCP_SLASH_ADD_PLAN_TOKEN_TTL_MS {
        return McpSlashAddPlanResult::Err {
            error: McpSlashAddPlanError::ExpiredToken {
                token: token.to_string(),
            },
        };
    }

    match add_mcp_config
        .add_config(&plan.server_name, &plan.config, plan.scope)
        .await
    {
        Ok(()) => McpSlashAddPlanResult::Ok { plan },
        Err(e) => McpSlashAddPlanResult::Err {
            error: McpSlashAddPlanError::InstallFailed {
                message: e.to_string(),
            },
        },
    }
}

/// Reset plan store for testing.
pub fn reset_mcp_slash_add_plan_store_for_testing() {
    PLAN_STORE.lock().unwrap().clear();
}
